src/Dialect/ONNX/ONNXOps/NN/Attention.cpp

```diff
@@ -0,0 +1,201 @@
+// (c) Copyright 2022 - 2025 Advanced Micro Devices, Inc. All Rights Reserved.
+
+// =============================================================================
+//
+// This file provides definition of ONNX dialect Attention operation.
+//
+//===----------------------------------------------------------------------===//
+
+#include "ONNXOps.hpp"
+
+#include "src/Dialect/ONNX/ONNXOps/OpHelper.hpp"
+
+using namespace mlir;
+using namespace mlir::OpTrait::util;
+using namespace onnx_mlir;
+
+namespace onnx_mlir {
+
+template <>
+LogicalResult ONNXAttentionOpShapeHelper::computeShape() {
+  auto attentionOp = cast<ONNXAttentionOp>(op);
+
+  const int64_t rank = createIE->getShapedTypeRank(attentionOp.getQ());
+  if (rank != 3 && rank != 4)
+    return failure();
+
+  DimsExpr qShape;
+  createIE->getShapeAsDims(attentionOp.getQ(), qShape);
+  DimsExpr kShape;
+  createIE->getShapeAsDims(attentionOp.getK(), kShape);
+  DimsExpr vShape;
+  createIE->getShapeAsDims(attentionOp.getV(), vShape);
+
+  auto qNumHeads = attentionOp.getQNumHeads();
+  auto kvNumHeads = attentionOp.getKvNumHeads();
+
+  auto normalizeInputTo4D = [](DimsExpr inputShape,
+                                std::optional<int64_t> numHeads) -> DimsExpr {
+    DimsExpr shape4D = inputShape;
+    if (inputShape.size() == 3) {
```

Comment:
tiny nit (can be ignore if you think it is too much trouble to do it for a small thing): invert the logic so that the if statement is a single line 🙂 
```suggestion
    if (inputShape.size() == 4) {
      return shape4D;
    }  
    assert(
          numHeads && "*_num_heads attributes must be present with 3D inputs");
    shape4D.insert(shape4D.begin() + 1, LitIE(*numHeads));
    shape4D[3] = shape4D[3].floorDiv(shape4D[1]);
```